README.txt

        mod_xmlrpc - XML-RPC server

        Homepage: http://www.ejabberd.im/mod_xmlrpc
        Author: Badlop

	Note: this module will only receive bugfixes.
	If you are using ejabberd trunk SVN 1635 or newer, you will probably
	prefer to install ejabberd_xmlrpc.
	

	DESCRIPTION
	-----------

mod_xmlrpc is a module for ejabberd, a XMPP/Jabber server written in Erlang.
It starts a XML-RPC server and waits for external requests. Implemented calls
include statistics and user administration. This allows external programs written
in any language like websites or administrative tools to communicate with
ejabberd to get information or to make changes without the need to know ejabberd
internals.

One example usage is a corporate site in PHP that creates a Jabber user every time
a new user is created on the website.

Some benefits of interfacing with the Jabber server by XML-RPC instead of modifying
directly the database are:
 - external programs are more simple and easy to develop and debug
 - can communicate with a server in a different machine, and even on Internet


	REQUIREMENTS
	------------

    ejabberd 1.1.2 or newer
    XMLRPC-Erlang 1.13 with IP, Ruby and Xmerl 1.x patches
    Optional: mod_muc_admin for MUC-related calls
    Optional: ejabberd SVN trunk r1329 or newer is required for muc_room_set_affiliation


 - Install XMLRPC-Erlang

wget http://ejabberd.jabber.ru/files/contributions/xmlrpc-1.13-ipr2.tgz
tar -xzvf xmlrpc-1.13-ipr2.tgz
cd xmlrpc-1.13/src
make
cd ../../



	CONFIGURE EJABBERD
	------------------

1. Add an option like this to the ejabberd start script:

$ erl -pa '/home/jabber/xmlrpc-1.13/ebin' ...

2. Configure ejabberd to start this module at startup:
edit ejabberd.cfg and add on 'modules' section:

	{mod_xmlrpc,      []},

3. Now start ejabberd. If everything went ok, no mention to mod_xmlrpc
will be shown on the error log.


	CONFIGURE
	---------

Configuration is done in 'modules' section in ejabberd.cfg. Available options:
    port: port where the XML-RPC server will listen. Default: 4560
    ip: IP address to listen, in Erlang format. Set to 'all' to listen on all IP address. Default: all
    maxsessions: Number of concurrent connections allowed. Default: 10
    timeout: Timeout of the connections, expressed in milliseconds. Default: 5000

Example configurations:
    {mod_xmlrpc,     [{port, 4560}]}

    {mod_xmlrpc,     [{port, 4560}, {ip, all}, {maxsessions, 10}, {timeout, 5000}]}

    {mod_xmlrpc,     [{port, 4560}, {ip, {81, 202, 202, 79}}]}



	USAGE
	-----

You can send calls to http://host:4560/

Call:           Arguments:                                                 Returns:

 -- debug
echothis        String                                                       String
multhis         struct[{a, Integer}, {b, Integer}]                          Integer

 -- statistics
tellme_title    String                                                       String
tellme_value    String                                                       String
tellme          String                     struct[{title, String}, {value. String}]

 -- user administration
create_account  struct[{user, String}, {host, String}, {password, String}]  Integer
delete_account  struct[{user, String}, {host, String}, {password, String}]  Integer
check_account   struct[{user, String}, {host, String}]                      Integer
check_password  struct[{user, String}, {host, String}, {password, String}]  Integer
change_password struct[{user, String}, {host, String}, {newpass, String}]   Integer
num_resources   struct[{user, String}, {host, String}]                      Integer
resource_num    struct[{user, String}, {host, String}, {num, Integer}]       String
set_nickname    struct[{user, String}, {host, String}, {nickname, String}]  Integer
add_rosteritem  struct[{localuser, String}, {localserver, String},
                       {user, String}, {server, String},
                       {nick, String}, {group, String}, {subs, String}]      String
delete_rosteritem  struct[{localuser, String}, {localserver, String},
                          {user, String}, {server, String}]                  String
get_roster  struct[{user, String}, {server, String}]
                      array[struct[{jid, String}, {nick, String}, {Group, String}]]
send_message  struct[{from, String}, {to, String}, {subject, String}, 
                      {body, String}]                                       Integer

 -- MUC administration
create_muc_room struct[{name, String}, {service, String}, {server, String}] Integer
delete_muc_room struct[{name, String}, {service, String}, {server, String}] Integer
muc_room_change_option struct[{name, String}, {service, String},
                              {option, String}, {value, String}]            Integer
muc_room_set_affiliation struct[{name, String}, {service, String},
                                {jid, String}, {affiliation, Affiliation}]  Integer

 -- Messaging
send_message struct[{jid, String}, {subject, String}, {body, String}]       Integer

	TEST
	----

 - You can easily try the XML-RPC server starting a new Erlang Virtual Machine
   and making calls to ejabberd's XML-RPC:

1. Start Erlang with this option:
$ erl -pa '/home/jabber/xmlrpc-1.13/ebin'

2. Now on the Erlang console, write commands and check the results:

> xmlrpc:call("127.0.0.1", 4560, "/", {call, create_muc_room, [{struct, [{name, "SiteMembers"}, {service, "conference.xmpp.darren.dev.spawareurope.com"}, {server, "xmpp.darren.dev.spawareurope.com"}]}]}).

1> xmlrpc:call({127, 0, 0, 1}, 4560, "/", {call, echothis, [800]}).
{ok,{response,[800]}}

2> xmlrpc:call({127, 0, 0, 1}, 4560, "/", {call, echothis, ["blot cloc 557.889 kg"]}).
{ok,{response,["blot cloc 557.889 kg"]}}

3> xmlrpc:call({127, 0, 0, 1}, 4560, "/", {call, multhis, [{struct,[{a, 83}, {b, 689}]}]}).
{ok,{response,[57187]}}

4> xmlrpc:call({127, 0, 0, 1}, 4560, "/", {call, create_account,
[{struct, [{user, "ggeo"}, {host, "example.com"}, {password, "gogo11"}]}]}).
{ok,{response,[0]}}

5> xmlrpc:call({127, 0, 0, 1}, 4560, "/", {call, create_account,
[{struct, [{user, "ggeo"}, {host, "example.com"}, {password, "gogo11"}]}]}).
{ok,{response,[409]}}

6> xmlrpc:call({127, 0, 0, 1}, 4560, "/", {call, muc_room_change_option,
[{struct, [{name, "test"}, {service, "conference.localhost"},
 {option, "title"}, {value, "Test Room"}]}]}).

7> xmlrpc:call({127, 0, 0, 1}, 4560, "/", {call, muc_room_set_affiliation,
[{struct, [{name, "test"}, {service, "conference.example.com"},
{jid, "ex@example.com"}, {affiliation, "member"}]}]}).

8> xmlrpc:call({127, 0, 0, 1}, 4560, "/", {call, muc_room_set_affiliation,
[{struct, [{name, "test"}, {service, "conference.example.com"},
{jid, "ex@example.com"}, {affiliation, "none"}]}]}).


 - Some possible XML-RPC error messages:

   + Client: connection refused: wrong IP, wrong port, the server is not started...

2> xmlrpc:call({127, 0, 0, 1}, 44444, "/", {call, echothis, [800]}).
{error,econnrefused}

   + Client: bad value: a800 is a string, so it must be put into ""

7> xmlrpc:call({127, 0, 0, 1}, 4560, "/", {call, echothis, [a800]}).
{error,{bad_value,a800}}

   + Server: unknown call: you sent a call that the server does not implement

3> xmlrpc:call({127, 0, 0, 1}, 4560, "/", {call, bububu, [800]}).
{ok,{response,{fault,-1,"Unknown call: {call,bububu,[800]}"}}}


	EXAMPLE IN PYTHON
	-----------------

This is an example XML-RPC client in Python, thanks to Diddek:
-------
import xmlrpclib

server_url = 'http://127.0.0.1:4560';
server = xmlrpclib.Server(server_url);

params = {}
params["user"] = "ggeo"
params["host"] = "localhost"
params["password"] = "gogo11"

result = server.create_account(params)
print result
-------


	EXAMPLE IN RUBY
	---------------

This is an example XML-RPC client in Ruby, thanks to Diddek:
-------
require 'xmlrpc/client'

host = "172.16.29.6:4560"
timeout = 3000000
client = XMLRPC::Client.new2("http://#{host}", "#{host}", timeout)
result = client.call("echothis", "800")
puts result
-------


	EXAMPLE IN PHP
	--------------

This is an XML-RPC client in PHP, thanks to Zbyszek Żółkiewski and Calder.
It requires "allow_url_fopen = On" in your php.ini.

-------
<?
$param=array("user"=>"test_user","host"=>"example.com","password"=>"some_password");
$request = xmlrpc_encode_request('check_password', $param, (array('encoding' => 'utf-8')));

$context = stream_context_create(array('http' => array(
    'method' => "POST",
    'header' => "User-Agent: XMLRPC::Client mod_xmlrpc\r\n" .
                "Content-Type: text/xml\r\n" .
                "Content-Length: ".strlen($request),
    'content' => $request
)));

$file = file_get_contents("http://127.0.0.1:4560/RPC2", false, $context);

$response = xmlrpc_decode($file);

if (xmlrpc_is_fault($response)) {
    trigger_error("xmlrpc: $response[faultString] ($response[faultCode])");
} else {
    print_r($response);
}

?>
-------


	EXAMPLE IN JAVA
	---------------

This is an XML-RPC client in Java, thanks to Calder.
It requires Apache XML-RPC available at http://ws.apache.org/xmlrpc/

-------
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

import org.apache.xmlrpc.client.XmlRpcClient;
import org.apache.xmlrpc.client.XmlRpcClientConfigImpl;

public class Test {

	public static void main(String[] args) {
		try {
		    XmlRpcClientConfigImpl config = new XmlRpcClientConfigImpl();
		    config.setServerURL(new URL("http://127.0.0.1:4560/RPC2"));
		    XmlRpcClient client = new XmlRpcClient();
		    client.setConfig(config);

		    /* Command string */
		    String command = "check_password";

		    /* Parameters as struct */
		    Map struct = new HashMap();
		    struct.put("user", "test1");
		    struct.put("host", "localhost");
		    struct.put("password", "test");

		    Object[] params = new Object[]{struct};
		    Integer result = (Integer) client.execute(command, params);
		    System.out.println(result);
		} catch (Exception e) {
			System.out.println(e);
		}
	}

}
-------


ChangeLog
2008-10-12  Badlop  <badlop@process-one.net>

	* README.txt: Add a note about ejabberd_xmlrpc existence for newer
	ejabberd trunk SVN 1635

2008-08-31  Badlop  <badlop@process-one.net>

	* README.txt: Added Java example client and fixed the PHP
	example (thanks to Calder)

2008-07-24  Badlop  <badlop@process-one.net>

	* src/mod_xmlrpc.erl: Fixed typo in get_roster

2008-07-08  Badlop  <badlop@process-one.net>

	* src/mod_xmlrpc.erl: Detect if mod_roster or mod_roster_odbc is
	enabled. New call send_message (thanks to Darren Ferguson)
	* README.txt: Likewise

2008-05-20  Badlop  <badlop@process-one.net>

	* src/mod_xmlrpc.erl: New call get_roster; works only with
	mod_roster_odbc
	* README.txt: Likewise

2008-05-19  Badlop  <badlop@process-one.net>

	* src/mod_xmlrpc.erl: New function check_account (thanks to
	Zbyszek Żółkiewski)
	* README.txt: Likewise

2008-05-18  Badlop  <badlop@process-one.net>

	* README.txt: Added example client in PHP (thanks to Zbyszek
	Żółkiewski)

	* src/mod_xmlrpc.erl: Convert from DOS to Unix line format
	* README.txt: Likewise
	* ChangeLog: Likewise

2008-05-16  Badlop  <badlop@process-one.net>

	* src/mod_xmlrpc.erl: New calls muc_room_change_option and
	muc_room_set_affiliation. Improved calls add_rosteritem and                     q
	delete_rosteritem: they push the changed roster item to connected
	resources (thanks to Darren Ferguson). New call check_password.
	* README.txt: Likewise

2008-05-06  Badlop  <badlop@process-one.net>

	* src/mod_xmlrpc.erl: Added new calls delete_account,
	delete_rosteritem, create_muc_room and destroy_muc_room (thanks to
	Darren Ferguson)
	* README.txt: Likewise

2008-02-29  Badlop  <badlop@process-one.net>

	* src/mod_xmlrpc.erl: Added module options: maxsessions, timeout
	* README.txt: Likewise

2007-08-24  Badlop  <badlop@ono.com>

	* README.txt: Removed requirement of Xmerl 0.20, since it is
	included in Erlang/OTP. Instead, require updated XMLRPC-Erlang.

2007-08-21  Badlop  <badlop@ono.com>

	* README.txt: New Ruby example (thanks to Diddek). Patched xmlrpc
	Erlang library for Ruby compatibility (thanks to Cstar).

	* mod_xmlrpc/*: Initial commit.

Old Changelog:

0.2.4 - 7 November 2006
  * Fixed a bug that required clients to provide attributes in a fixed order

0.2.3 - 22 September 2006
  * New feature: bind to a specific IP address, requires the patched XML-RPC-Erlang-1.13-IP (thanks to Zeank)

0.2.2 - 15 August 2006
  * Fixed small bug on resource_num (thanks to Flipkick)

0.2.1 - 20 July 2006
  * Fixed small bug on add_rosteritem (thanks to Leonexis)

0.2.0 - 16 April 2006
  * Added two new calls: num_resources and resouce_num
  * Added support for vhosts

0.1.2 - 28 December 2005
  * Now compatible with ejabberd 1.0.0
  * The XMLRPC server is started only once, not once for every virtual host
  * Added comments for handlers. Every available handler must be explained

0.1.0 - 4 April 2005
  * Initial version


src/mod_xmlrpc.erl
%%%----------------------------------------------------------------------
%%% File    : mod_xmlrpc.erl
%%% Author  : Badlop <badlop@ono.com>
%%% Purpose : XML-RPC server
%%% Created : 21 Aug 2007 by Badlop <badlop@ono.com>
%%% Id      : $Id: mod_xmlrpc.erl 1061 2010-02-19 14:17:13Z badlop $
%%%----------------------------------------------------------------------

-module(mod_xmlrpc).
-author('badlop@ono.com').

-behaviour(gen_mod).

-export([start/2,
	 handler/2,
	 loop/1,
	 stop/1]).

-include("ejabberd.hrl").
-include("mod_roster.hrl").
-include("jlib.hrl").

-define(PROCNAME, ejabberd_mod_xmlrpc).

%% -----------------------------
%% Module interface
%% -----------------------------

start(_Host, Opts) ->
    case whereis(?PROCNAME) of
	undefined ->
	    %% get options
	    Port = gen_mod:get_opt(port, Opts, 4560),
	    MaxSessions = gen_mod:get_opt(maxsessions, Opts, 10),
	    Timeout = gen_mod:get_opt(timeout, Opts, 5000),
	    Handler = {mod_xmlrpc, handler},
	    State = started,

	    Ip = gen_mod:get_opt(ip, Opts, all),

	    %% start the XML-RPC server
	    {ok, Pid} = xmlrpc:start_link(Ip, Port, MaxSessions, Timeout, Handler, State),

	    %% start the loop process
	    register(?PROCNAME, spawn(?MODULE, loop, [Pid]));
	_ ->
	    ok
    end.

loop(Pid) ->
    receive
        stop ->
	    xmlrpc:stop(Pid)
    end.

stop(_Host) ->
    ?PROCNAME ! stop.


%% -----------------------------
%% Handlers
%% -----------------------------

%% Call:           Arguments:                                      Returns:

%% .............................
%%  Debug

%% echothis        String                                          String
handler(_State, {call, echothis, [A]}) ->
    {false, {response, [A]}};

%% multhis         struct[{a, Integer}, {b, Integer}]              Integer
handler(_State, {call, multhis, [{struct, [{a, A}, {b, B}]}]}) ->
    {false, {response, [A*B]}};

%% .............................
%%  Statistics

%% tellme_title    String                                          String
handler(_State, {call, tellme_title, [A]}) ->
    {false, {response, [get_title(A)]}};

%% tellme_value    String                                          String
handler(_State, {call, tellme_value, [A]}) ->
    N = node(),
    {false, {response, [get_value(N, A)]}};

%% tellme          String          struct[{title, String}, {value. String}]
handler(_State, {call, tellme, [A]}) ->
    N = node(),
    T = {title, get_title(A)},
    V = {value, get_value(N, A)},
    R = {struct, [T, V]},
    {false, {response, [R]}};

%% .............................
%%  User administration

%% create_account  struct[{user, String}, {host, String}, {password, String}]      Integer
handler(_State, {call, create_account, [{struct, AttrL}]}) ->
    [U, H, P] = get_attrs([user, host, password], AttrL),
    R = case jlib:nodeprep(U) of
	    error ->
		error;
	    "" ->
		error;
	    _ ->
		ejabberd_auth:try_register(U, H, P)
	end,
    case R of
	{atomic, ok} ->
	    {false, {response, [0]}};
	{atomic, exists} ->
	    {false, {response, [409]}};
	_E ->
	    {false, {response, [1]}}
    end;

%% delete_account  struct[{user, String}, {host, String}, {password, String}]      Integer
handler(_State, {call, delete_account, [{struct, AttrL}]}) ->
    [U, H, P] = get_attrs([user, host, password], AttrL),
    case jlib:nodeprep(U) of
        error ->
	    error;
        "" ->
	    error;
        _ ->
	    ejabberd_auth:remove_user(U, H, P)
    end,
    {false, {response, [0]}};

%% check_account  struct[{user, String}, {host, String}]                           Integer
handler(_State, {call, check_account, [{struct, AttrL}]}) ->
    [U, H] = get_attrs([user, host], AttrL),
    case ejabberd_auth:is_user_exists(U, H) of
	true ->
	    {false, {response, [0]}};
	false ->
	    {false, {response, [1]}}
    end;

%% check_password struct[{user, String}, {host, String}, {password, String}]       Integer
handler(_State, {call, check_password, [{struct, AttrL}]}) ->
    [U, H, P] = get_attrs([user, host, password], AttrL),
    case ejabberd_auth:check_password(U, H, P) of
	true ->
	    {false, {response, [0]}};
	false ->
	    {false, {response, [1]}}
    end;

%% change_password struct[{user, String}, {host, String}, {newpass, String}]       Integer
handler(_State, {call, change_password, [{struct, AttrL}]}) ->
    [U, H, P] = get_attrs([user, host, newpass], AttrL),
    case ejabberd_auth:set_password(U, H, P) of
	ok ->
	    {false, {response, [0]}};
	_ ->
	    {false, {response, [1]}}
    end;

%% num_resources struct[{user, String}, {host, String}]                            Integer
handler(_State, {call, num_resources, [{struct, AttrL}]}) ->
    [U, H] = get_attrs([user, host], AttrL),
    R = length(ejabberd_sm:get_user_resources(U, H)),
    {false, {response, [R]}};

%% resource_num struct[{user, String}, {host, String}, {num, Integer}]             String
handler(_State, {call, resource_num, [{struct, AttrL}]}) ->
    [U, H, N] = get_attrs([user, host, num], AttrL),
    Resources = ejabberd_sm:get_user_resources(U, H),
    case (0<N) and (N=<length(Resources)) of
	true ->
	    R = lists:nth(N, Resources),
	    {false, {response, [R]}};
	false ->
	    FaultString = lists:flatten(io_lib:format("Wrong resource number: ~p", [N])),
	    {false, {response, {fault, -1, FaultString}}}
    end;

%%% set_nickname    struct[{user, String}, {host, String}, {nickname, String}]      Integer
handler(_State, {call, set_nickname, [{struct, AttrL}]}) ->
    [U, H, N] = get_attrs([user, host, nickname], AttrL),
    R = mod_vcard:process_sm_iq(
	  {jid, U, H, "", U, H, ""},
	  {jid, U, H, "", U, H, ""},
	  {iq, "", set, "", "en",
	   {xmlelement, "vCard",
	    [{"xmlns", "vcard-temp"}], [
					{xmlelement, "NICKNAME", [], [{xmlcdata, N}]}
				       ]
	   }}),
    case R of
	{iq, [], result, [], _L, []} ->
	    {false, {response, [0]}};
	_ ->
	    {false, {response, [1]}}
    end;

%% add_rosteritem  struct[{localuser, String}, {localserver, String},
%%                  {user, String}, {server, String},
%%                  {nick, String}, {group, String},
%%                  {subs, String}]                                String
handler(_State, {call, add_rosteritem, [{struct, AttrL}]}) ->
    [LocalUser, LocalServer, User, Server, Nick, Group, Subs] =
	get_attrs([localuser, localserver, user, server, nick, group, subs], AttrL),
    Node = node(),
    R = case add_rosteritem(LocalUser, LocalServer, User, Server, Nick, Group, list_to_atom(Subs), []) of
	    {atomic, ok} ->
                push_roster_item(LocalUser, LocalServer, User, Server, {add, Nick, Subs, Group}),
		0;
	    {error, Reason} ->
		io:format("Can't add roster item to user ~p@~p on node ~p: ~p~n",
			  [LocalUser, LocalServer, Node, Reason]),
		1;
	    {badrpc, Reason} ->
		io:format("Can't add roster item to user ~p@~p on node ~p: ~p~n",
			  [LocalUser, LocalServer, Node, Reason]),
		1
	end,
    {false, {response, [R]}};


%% delete_rosteritem  struct[{localuser, String}, {localserver, String},
%%                     {user, String}, {server, String}]            String
handler(_State, {call, delete_rosteritem, [{struct, AttrL}]}) ->
    [LocalUser, LocalServer, User, Server] = get_attrs([localuser, localserver, user, server], AttrL),
    Node = node(),
    R = case delete_rosteritem(LocalUser, LocalServer, User, Server) of
            {atomic, ok} ->
                push_roster_item(LocalUser, LocalServer, User, Server, remove),
                0;
            {error, Reason} ->
                io:format("Can't remove roster item from user ~p@~p on node ~p: ~p~n",
			  [LocalUser, LocalServer, Node, Reason]),
		1;
	    {badrpc, Reason} ->
		io:format("Can't remove roster item from user ~p@~p on node ~p: ~p~n",
			  [LocalUser, LocalServer, Node, Reason]),
		1
	end,
    {false, {response, [R]}};

%% get_roster  struct[{user, String}, {server, String}]
%%                       array[struct[{jid, String}, {nick, String}, {Group, String}]]
handler(_State, {call, get_roster, [{struct, AttrL}]}) ->
    [User, Server] = get_attrs([user, server], AttrL),
    Node = node(),
    R = case get_roster(User, Server) of
	    {ok, Roster} ->
                RosterXMLRPC = make_roster_xmlrpc(Roster),
		{array, RosterXMLRPC};
	    {error, Reason} ->
		io:format("Can't get roster of user ~p@~p on node ~p: ~p~n",
			  [User, Server, Node, Reason]),
		1;
	    {badrpc, Reason} ->
		io:format("Can't get roster of user ~p@~p on node ~p: ~p~n",
			  [User, Server, Node, Reason]),
		1
	end,
    {false, {response, [R]}};

%% create_muc_room  struct[{name, String}, {service, String}, {server, String}]  Integer
handler(_State, {call, create_muc_room, [{struct, AttrL}]}) ->
    [Name, Service, Server] = get_attrs([name, service, server], AttrL),
    R = case mod_muc_admin:create_room(Name, Service, Server) of
	    ok ->
		0;
	    {error, room_already_exists} ->
		1
	end,
    {false, {response, [R]}};

%% delete_muc_room  struct[{name, String}, {service, String}, {server, String}]  Integer
handler(_State, {call, delete_muc_room, [{struct, AttrL}]}) ->
    [Name, Service, Server] = get_attrs([name, service, server], AttrL),
    R = case mod_muc_admin:destroy_room(Name, Service, Server) of
	    ok ->
		0;
	    {error, room_not_exists} ->
		?INFO_MSG("~n MUC: ~p does not exist on service ~p ~n", [Name, Service]),
		1
	end,
    {false, {response, [R]}};

%% muc_room_change_option  struct[{name, String}, {service, String}, {option, String}, {value String}]  Integer
handler(_State, {call, muc_room_change_option, [{struct, AttrL}]}) ->
    [Name, Service, OptionString, Value] = get_attrs([name, service, option, value], AttrL),
    Option = list_to_atom(OptionString),
    mod_muc_admin:change_room_option(Name, Service, Option, Value),
    {false, {response, [0]}};

%% muc_room_set_affiliation  struct[{name, String}, {service, String}, {jid, String}, {affiliation, String}] Integer
handler(_State, {call, muc_room_set_affiliation, [{struct, AttrL}]}) ->
    [Name, Service, JID, Affiliation] = get_attrs([name, service, jid, affiliation], AttrL),
    R = case mod_muc_admin:set_affiliation(Name, Service, JID, list_to_atom(Affiliation)) of
	    ok ->
		0;
	    {error, room_not_exists} ->
		?INFO_MSG("~n MUC: ~p does not exist on service ~p ~n", [Name, Service]),
		1
	end,
    {false, {response, [R]}};

%% send_message  struct[{from, String}, {to, String}, {subject, String}, {body, String}]  Integer
handler(_State, {call, send_message, [{struct, AttrL}]}) ->
    [FromJIDString, ToJIDString, Subject, Body] = get_attrs([from, to, subject, body], AttrL),
    FromJID = jlib:string_to_jid(FromJIDString),
    ToJID = jlib:string_to_jid(ToJIDString),
    send_message(FromJID, ToJID, Subject, Body),
    {false, {response, [0]}};

%% set_presence struct[{user, String}, {server, String}, {show, String}, {status, String}, {type, String}] Integer
handler(_State, {call, set_presence, [{struct, AttrL}]}) ->
    [User, Server, Show, Status, Type] = get_attrs([user, server, show, status, type], AttrL),
    JID = jlib:make_jid(User, Server, ""),
    if
       Type == "unavailable" ->
          MPacket = set_user_presence_unavailable(Type, Show, Status);
       true ->
          MPacket = set_user_presence_available(Show, Status)
    end,
    R = case get_roster(User, Server) of
        {ok, Roster} ->
            _ = send_user_presence_packets(Roster, MPacket, JID),
	    0;
	{error, Reason} ->
	    io:format("Can't get roster of user ~p@~p: reason: ~p~n", [User, Server, Reason]),
	    1;
	{badrpc, Reason} ->
	    io:format("Can't get roster of user ~p@~p: ~p~n", [User, Server, Reason]),
	    1
    end,
    {false, {response, [R]}};

%% If no other guard matches
handler(_State, Payload) ->
    FaultString = lists:flatten(io_lib:format("Unknown call: ~p", [Payload])),
    ?INFO_MSG("Unknown call: ~p", [Payload]),
    {false, {response, {fault, -1, FaultString}}}.


%% -----------------------------
%% Roster items
%% -----------------------------

add_rosteritem(LU, LS, User, Server, Nick, Group, Subscription, Xattrs) ->
    subscribe(LU, LS, User, Server, Nick, Group, Subscription, Xattrs).

subscribe(LU, LS, User, Server, Nick, Group, Subscription, Xattrs) ->
    mnesia:transaction(
      fun() ->
	      mnesia:write({roster,
			    {LU,LS,{User,Server,[]}}, % uj
			    {LU,LS},                  % user
			    {User,Server,[]},      % jid
			    Nick,                  % name: "Mom", []
			    Subscription,  % subscription: none, to=you see him, from=he sees you, both
			    none,          % ask: out=send request, in=somebody requests you, none
			    [Group],       % groups: ["Family"]
			    Xattrs,        % xattrs: [{"category","conference"}]
			    []             % xs: []
			   })
      end).

delete_rosteritem(LU, LS, User, Server) ->
    unsubscribe(LU, LS, User, Server).

unsubscribe(LU, LS, User, Server) ->
    mnesia:transaction(
      fun() ->
              mnesia:delete({roster, {LU, LS, {User, Server, []}}})
      end).

%% @spec(LU, LS, U, S, Action) -> ok
%%       Action = {add, Nick, Subs, Group} | remove
%% @doc Push to the roster of account LU@LS the contact U@S.
%% The specific action to perform is defined in Action.
push_roster_item(LU, LS, U, S, Action) ->
    lists:foreach(fun(R) ->
			  push_roster_item(LU, LS, R, U, S, Action)
		  end, ejabberd_sm:get_user_resources(LU, LS)).

push_roster_item(LU, LS, R, U, S, Action) ->
    Item = build_roster_item(U, S, Action),
    ResIQ = build_iq_roster_push(Item),
    LJID = jlib:make_jid(LU, LS, R),
    ejabberd_router:route(LJID, LJID, ResIQ).

build_roster_item(U, S, {add, Nick, Subs, Group}) ->
    {xmlelement, "item",
     [{"jid", jlib:jid_to_string(jlib:make_jid(U, S, ""))},
      {"name", Nick},
      {"subscription", Subs}],
     [{xmlelement, "group", [], [{xmlcdata, Group}]}]
    };
build_roster_item(U, S, remove) ->
    {xmlelement, "item",
     [{"jid", jlib:jid_to_string(jlib:make_jid(U, S, ""))},
      {"subscription", "remove"}],
     []
    }.

build_iq_roster_push(Item) ->
    {xmlelement, "iq",
     [{"type", "set"}, {"id", "push"}],
     [{xmlelement, "query",
       [{"xmlns", ?NS_ROSTER}],
       [Item]
      }
     ]
    }.


%% -----------------------------
%% Get Roster
%% -----------------------------

get_roster(User, Server) ->
    LUser = jlib:nodeprep(User),
    LServer = jlib:nameprep(Server),
    Roster = ejabberd_hooks:run_fold(roster_get, LServer, [], [{LUser, LServer}]),
    {ok, Roster}.

%% Note: if a contact is in several groups, the contact is returned
%% several times, each one in a different group.
make_roster_xmlrpc(Roster) ->
    lists:foldl(
      fun(Item, Res) ->
	      JIDS = jlib:jid_to_string(Item#roster.jid),
	      Nick = Item#roster.name,
	      Groups = case Item#roster.groups of
			   [] -> [""];
			   Gs -> Gs
		       end,
	      ItemsX = [{struct, [{jid, JIDS}, {nick, Nick}, {group, Group}]}
			|| Group <- Groups],
	      ItemsX ++ Res
      end,
      [],
      Roster).

%% -----------------------------
%% Sending messages to the user
%% -----------------------------

%% @doc Send a message to a Jabber account.
%% If a resource was specified in the JID,
%% the message is sent only to that specific resource.
%% If no resource was specified in the JID,
%% and the user is remote or local but offline,
%% the message is sent to the bare JID.
%% If the user is local and is online in several resources,
%% the message is sent to all its resources.
send_message(FromJID, ToJID, Subject, Body) ->
    ToUser = ToJID#jid.user,
    ToServer = ToJID#jid.server,
    case ToJID#jid.resource of
	"" ->
	    send_message(FromJID, ToUser, ToServer, Subject, Body);
	Resource ->
	    send_message(FromJID, ToUser, ToServer, Resource, Subject, Body)
    end.

send_message(FromJID, ToUser, ToServer, Subject, Body) ->
    case ejabberd_sm:get_user_resources(ToUser, ToServer) of
	[] ->
	    send_message(FromJID, ToUser, ToServer, "", Subject, Body);
	ToResources ->
	    lists:foreach(
	      fun(ToResource) ->
		      send_message(FromJID, ToUser, ToServer, ToResource, Subject, Body)
	      end,
	      ToResources)
    end.

send_message(FromJID, ToU, ToS, ToR, Subject, Body) ->
    MPacket = build_send_message(Subject, Body),
    ToJID = jlib:make_jid(ToU, ToS, ToR),
    ejabberd_router:route(FromJID, ToJID, MPacket).

build_send_message(Subject, Body) ->
    {xmlelement, "message",
     [{"type", "headline"}],
     [{xmlelement, "subject", [], [{xmlcdata, Subject}]},
      {xmlelement, "body", [], [{xmlcdata, Body}]}
     ]
    }.

%% -----------------------------
%% Set Presence
%% -----------------------------
set_user_presence_available(Show, Status) ->
    {xmlelement, "presence",
     [],
     [{xmlelement, "show", [], [{xmlcdata, Show}]},
      {xmlelement, "status", [], [{xmlcdata, Status}]}
     ]
    }.

set_user_presence_unavailable(Type, Show, Status) ->
    {xmlelement, "presence",
     [{"type", Type}],
     [{xmlelement, "show", [], [{xmlcdata, Show}]},
      {xmlelement, "status", [], [{xmlcdata, Status}]}
     ]
    }.

%% Sending the presence to each user in your roster
send_user_presence_packets(Roster, Packet, JID) ->
    lists:foldl(
      fun(Item, _) ->
         JIDS = jlib:jid_to_string(Item#roster.jid),
         SJID = jlib:string_to_jid(JIDS),
         ejabberd_router:route(JID, SJID, Packet)
      end,
      [],
      Roster).

%% -----------------------------
%% Internal
%% -----------------------------

get_title(A) -> mod_statsdx:get_title(A).
get_value(N, A) -> mod_statsdx:get(N, [A]).

get_attrs(Attribute_names, L) ->
    [get_attr(A, L) || A <- Attribute_names].
get_attr(A, L) ->
    case lists:keysearch(A, 1, L) of
	{value, {A, Value}} -> Value;
	false ->
	    %% Report the error and then force a crash
	    ?ERROR_MSG("Attribute '~p' not found on the list of attributes provided on the call:~n ~p", [A, L]),
	    attribute_not_found = A
    end.

